### PR TITLE
Fix _xla_send_recv_validation in collective pipeliner

### PR DESCRIPTION
**Pull request by shraiysh**
This patch fixes the value of `_xla_send_recv_validation` attribute in
collective-pipeliner. Collective pipeliner peels one iteration outside
loop and updates the initialization variable of the loop. This means
that the trip count of the loop reduces.

More details about this are added as docstrings.

### CODE REVIEW

**Review by frgossen — CHANGES_REQUESTED** [review]
How does this relate to your previous PR 12892? What aspect does it fix?

File: xla/service/collective_pipeliner.cc [hunk 1]
```diff
@@ -1197,6 +1199,94 @@ HloInstruction* CreateZero(HloComputation* comp, const Shape& shape,
...
+      send_recv_validation_attr = std::vector<Interval>(n_pairs, {1, 0});
+      send_recv_validation_attr.front().first = 0;
+    }
+  } else {
+    return absl::UnimplementedError(
+        "Only forward and backward pipelining are handled for the "
+        "_xla_send_recv_validation attribute for collective permutes");
+  }
+
+  std::string send_recv_validation_attr_str =
```
Comment (frgossen) [inline xla/service/collective_pipeliner.cc:1272]: This appears in multiple PRs. Can you extract it as a function and reuse it?

**Review by frgossen — CHANGES_REQUESTED** [review]

**Comment by shraiysh** [comment]
> How does this relate to your previous PR 12892? What aspect does it fix?

This is not about fixing a logical issue in the code. This fixes the value of the attribute when collective pipeliner is run. Collective pipeliner peels one iteration, where the peeled collective ends up before (or after) the loop if the pipelining direction is backward (or forward). This means that there is one less iteration and so the "valid iterations" should reflect that change. The peeled collective should also have a proper annotation indicating which device should run it.

On a related note, can you add @bixia1 as a review too? I think it would be good if she could check the testcases and give a go ahead for this fix.

**Comment by shraiysh** [comment]
I think I need to fix the logic in this a bit. Converting to draft for sometime while I figure this out.

**Comment by shraiysh** [comment]
This is ready for review now @bixia1 @frgossen

**Review by frgossen — APPROVED** [review]
Thanks!

**Comment by frgossen** [comment]
This needs another rebase to pass the internal CI

**Comment by shraiysh** [comment]
>This needs another rebase to pass the internal CI

Done. Thanks for the review!

**Review by bixia1 — APPROVED** [review]

**Comment by shraiysh** [comment]
@frgossen any ideas why this isn't going through? The CI build failure seems unrelated.

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
It looks like somehow the approval of @bixia1 didn't count. I approved again.

**Pull request closed without merging**